tests/foreman/cli/test_variables.py

```diff
@@ -0,0 +1,1120 @@
+        """Create Smart Variable with an existing name.
+
+        @id: 7f37194d-4a12-437b-a284-3350cf048eea
+
+        @steps:
+
+        1. Create a smart Variable with Valid name and default value.
+        2. Attempt to create a variable with same name from same/other class.
+
+        @assert: The variable with same name are not allowed to create from
+        any class.
+
+        @caseautomation: notautomated
+
+        @CaseLevel: System
+        """
+
+    @run_only_on('sat')
+    @stubbed()
+    @tier1
+    def test_positive_create_type(self):
+        """Create smart variable with all valid key types and values.
+
+        Types - string, boolean, integer, real, array, hash, yaml, json
+
+        @id: d89cdd32-dd2b-46fe-bcc2-8c66d92cc6f8
+
+        @steps:
+
+        1.  Create a variable with valid key type and default value.
+
+        @assert: Variable is created with a new type successfully.
+
+        @caseautomation: notautomated
+        """
+
+    @run_only_on('sat')
+    @stubbed()
+    @tier1
+    def test_negative_update_type(self):
+        """Create smart variable with all valid key types but invalid values.
+
+        Types - string, boolean, integer, real, array, hash, yaml, json
+
+        @id: d64d1b6d-028a-4782-a6d4-e3029d7118b6
+
+        @steps:
+
+        1.  Create a variable with valid key type and invalid default value.
+
+        @assert: Variable is not created with new type for invalid value.
+
+        @caseautomation: notautomated
+        """
+
+    @run_only_on('sat')
+    @stubbed()
+    @tier1
+    def test_negative_validate_default_value_required_flag(self):
+        """Test required flag works for empty default value.
+
+        @id: cdbe5feb-daf1-49c4-98c0-5bbbd436c3e8
+
+        @steps:
+
+        1.  Create a variable with valid default value and required flag to
+        True.
+        2. Attempt to override the variable with empty value from associated
+        host.
+
+        @assert: The variable value is not overridden with empty value.
+
+        @caseautomation: notautomated
+        """
+
+    @run_only_on('sat')
+    @stubbed()
+    @tier1
+    def test_positive_validate_default_value_required_flag(self):
+        """Test required flag works for non empty default value.
+
+        @id: 09e42df9-2fe1-444f-a603-6f7b6b3f127e
+
+        @steps:
+
+        1.  Create a variable with some default value and required flag set to
+        True.
+        2. Attempt to override the variable with valid value from associated
+        host.
+
+        @assert: The variable value is overridden with empty value.
+
+        @caseautomation: notautomated
+        """
+
+    @run_only_on('sat')
+    @stubbed()
+    @tier1
+    def test_negative_validate_matcher_value_required_flag(self):
+        """Test required flag works for empty matcher value.
+
+        @id: 7e02efb6-e1a2-4872-877a-78075b2de182
+
+        @steps:
+
+        1.  Create a matcher for variable for some attribute with blank value.
```

Comment:
I don't think that we can create empty matcher no matter we have flag specified or not. Please, check that